### ISSUE TITLE
2/5 missing presentation frames (very weird)

### ISSUE THREAD

**Issue by Radamanthe** [LunarG/VulkanSamples#83]
Hi,

This is an issue I have on my Debian system. I'm quite experienced in game engine programming, but I've never seen that before.

It's simple to describe: the presentation works 3 times out of 5. Period. Whatever I do. But let me develop a bit: when calling vkQueuePresent(), here is what happens for each frame:

- frame 1 is presented
- frame 2 is NOT presented (still see frame 1)
- frame 3 is NOT presented (still see frame 1)
- frame 4 is presented
- frame 5 is presented
- frame 6 is presented
- frame 7 is NOT presented (still see frame 6)
- frame 8 is NOT presented (still see frame 6)
- frame 9 is presented
- frame 10 is presented
- frame 11 is presented
- frame 12 is NOT presented (still see frame 11)
... (and so on indefinitely)

What is weird about it is that it's not the program fault at all. Proof? Well, I did this patch on demo.c sample to make sure:

```
From 9685bafb791968502f45e243a91dbef1f906c503 Mon Sep 17 00:00:00 2001
From: Alexandre BACQUART
Date: Fri, 13 May 2016 23:37:02 +0200
Subject: [PATCH] test driver (maybe) issue

---
 demos/cube.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/demos/cube.c b/demos/cube.c
index bc39141..2226137 100644
--- a/demos/cube.c
+++ b/demos/cube.c
@@ -30,6 +30,7 @@
 #include <signal.h>
 #ifdef __linux__
 #include <X11/Xutil.h>
+#include <unistd.h>
 #endif
 
 #ifdef _WIN32
@@ -779,6 +780,10 @@ static void demo_draw(struct demo *demo) {
         assert(!err);
     }
 
+    static unsigned frameCount = 0;
+    printf("frame %u\n", ++frameCount);
+    usleep(500000);
+
     err = vkQueueWaitIdle(demo->queue);
     assert(err == VK_SUCCESS);
 
@@ -2982,7 +2987,7 @@ static void demo_init(struct demo *demo, int argc, char **argv) {
     demo->width = 500;
     demo->height = 500;
 
-    demo->spin_angle = 0.01f;
+    demo->spin_angle = 2.f;
     demo->spin_increment = 0.01f;
     demo->pause = false;
 
-- 
2.8.0.rc3

```

This simple patch idea is to make the issue obvious (at least to me, because TBH, I would be surprised that this behavior happens for everyone, even considering only GNU/Linux. If it does the same for you, please let me know). The patch does 3 things:

- Add a sleep period of 500ms after each call to vkQueuePresent(), this way I can clearly see every frame (note that changing this value DOES NOT change the behavior described above).
- Add a frame counter output to stdout to follow it more closely (launcher terminal on a side, the cube sample window on the other side).
- Set the initial spin angle to 2 to exacerbate the difference between frames.

And the result is exactly what I described. Of course, without the introduced delay, this issue is not noticeable in IMMEDIATE presentation mode, and barely noticeable in FIFO presentation mode (though it's not as smooth as it should be, it's hard to guess what is happening).

Also note that validation layers are completely silent. Nothing seems to be wrong from the Vulkan implementation POV. Another note: my OpenGL based engine does not have this issue, all frames are presented and everything is perfectly smooth.

My system is Debian (sid). Installed drivers are the proprietary NVidia 364.19 (GPU is GeForce GTX 660). CPU is a quite old dual-core AMD Athlon 64, though I highly doubt it is relevant anyway.

Now, I highly suspect a driver issue, but I'm not 100% certain. Maybe it's not the right place to talk about that. So don't hesitate to close this issue, but please tell me what you think about it (and maybe where I could discuss about it).

Thanks.

**Comment by ianelliottus** [comment]
Hi @Radamanthe,

I just applied your "cube.c" changes and tried it on my Ubuntu system, also with NVIDIA graphics (sorry, I don't know which card), and saw some anomalies.  Normally, cube looks pretty steady for me, but it was good to slow it down, per your changes.  What I saw was:

- Occasional "stutter".  That is, most of the time the cube rotated in a certain direction, about 2FPS.  Sometimes, it would go back-and-forth very quickly (i.e. go forwards and backwards at a must faster frame rate) for one of the frames.

- A few times, I didn't see a new frame drawn (like what you reported, only much more occasionally).

I then ported this to Windows, and ran it on a laptop with an Intel Skylake GPU.  I saw no visual anomalies at all.  My guess is that this is a driver bug.

**Comment by Radamanthe** [comment]
Yes, by default (without modification), the cube.c sample is quite steady for me too. But that's not surprising: in IMMEDIATE presentation mode, it runs at several thousands FPS! In other words, the presentation is made several times per V-Blank (maybe like ~20 or more times). So we cannot see that some of them are missing. It's a little more obvious when simply changing the presentation mode to FIFO, logically limiting the number of presented frames to 1 per V-Blank.

I made this patch on cube.c because that behavior I described happened to my current Vulkan engine (WIP) and I wanted to make sure I did not do something wrong. Your description is quite different. Mine is **always exactly** as I described it. Yours seems a bit more random (still erratic though).

Anyway, I'm quite sure it's a driver bug. BTW, it's not very surprising that the GNU/Linux version of the drivers is more bugged than Wndows version. So I can only guess it will be fixed at some point. I think I'll take some time soon to report it properly to NVidia directly. Thanks for sharing your experience.

Even though it doesn't look like a LunarG issue, I won't close it for now. Again, I would not take it wrong if LunarG team decides to close it promptly. I just hope some more people will try it on their GNU/Linux rig and share their experience.
